src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributeList.cs

```diff
@@ -0,0 +1,231 @@
+        public void Add([NotNull] string name, object value)
+        {
+            Attributes.Add(new TagHelperAttribute(name, value));
+        }
+
+        /// <inheritdoc />
+        /// <remarks>
+        /// <paramref name="attribute"/>'s <see cref="TagHelperAttribute.Name"/> must not be <c>null</c>.
+        /// </remarks>
+        public void Add([NotNull] TagHelperAttribute attribute)
+        {
+            ValidateAttribute(attribute, nameof(attribute));
+
+            Attributes.Add(attribute);
+        }
+
+        /// <inheritdoc />
+        /// <remarks>
+        /// <paramref name="attribute"/>'s <see cref="TagHelperAttribute.Name"/> must not be <c>null</c>.
+        /// </remarks>
+        public void Insert(int index, [NotNull] TagHelperAttribute attribute)
+        {
+            ValidateAttribute(attribute, nameof(attribute));
+
+            Attributes.Insert(index, attribute);
+        }
+
+        /// <inheritdoc />
+        public void CopyTo([NotNull] TagHelperAttribute[] array, int index)
+        {
+            Attributes.CopyTo(array, index);
+        }
+
+        /// <inheritdoc />
+        /// <remarks>
+        /// <paramref name="attribute"/>s <see cref="TagHelperAttribute.Name"/> is compared case-insensitively.
+        /// </remarks>
+        public bool Remove([NotNull] TagHelperAttribute attribute)
+        {
+            return Attributes.Remove(attribute);
+        }
+
+        /// <inheritdoc />
+        public void RemoveAt(int index)
+        {
+            Attributes.RemoveAt(index);
+        }
+
+        /// <summary>
+        /// Removes all <see cref="TagHelperAttribute"/>s with <see cref="TagHelperAttribute.Name"/> matching
+        /// <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">
+        /// The <see cref="TagHelperAttribute.Name"/> of <see cref="TagHelperAttribute"/>s to remove.
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if at least 1 <see cref="TagHelperAttribute"/> was removed; otherwise, <c>false</c>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public bool RemoveAll([NotNull] string name)
+        {
+            return Attributes.RemoveAll(attribute => NameEquals(name, attribute)) > 0;
+        }
+
+        /// <inheritdoc />
+        public void Clear()
+        {
+            Attributes.Clear();
+        }
+
+        private void ValidateAttribute(TagHelperAttribute attribute, string parameterName)
+        {
+            if (attribute.Name == null)
+            {
+                throw new ArgumentException(
+                    Resources.FormatTagHelperAttributeList_CannotAddWithNullName(
+                        typeof(TagHelperAttribute).FullName,
+                        nameof(TagHelperAttribute.Name)),
+                    parameterName);
+            }
```

Comment:
only a few lines and the method makes debugging more annoying than it already is.  suggest in-lining into the 3 call sites.